core/src/autogluon/core/scheduler/jobs.py

```diff
@@ -0,0 +1,157 @@
+import logging
+import multiprocessing as mp
+import os
+import random
+import sys
+import time
+from functools import partial
+from time import sleep
+
+import dill
+
+from autogluon.core import make_temp_directory, Task
+from autogluon.core.scheduler.managers import TaskManagers
+from autogluon.core.scheduler.reporter import Communicator, LocalStatusReporter
+from autogluon.core.utils import CustomProcess, AutoGluonEarlyStop
+from autogluon.core.utils.multiprocessing_utils import is_fork_enabled
+import threading
+
+logger = logging.getLogger(__name__)
+
+__all__ = ['DistributedJobRunner']
+
+
+class DistributedJobRunner(object):
+
+    JOB_REGISTRATION_LOCK = mp.Lock()
+
+    def __init__(self, task: Task, managers: TaskManagers):
+        self.task: Task = task
+        self.managers: TaskManagers = managers
+
+    def start_distributed_job(self):
+        """Async Execute the job in remote and release the resources
+        """
+        logger.debug(f'Scheduling {self.task}')
+
+        def _release_resource_callback(fut):
+            logger.debug(f'Releasing Resources {self.task.task_id}')
+            with DistributedJobRunner.JOB_REGISTRATION_LOCK:
+                self.managers.release_resources(self.task.resources)
+
+        with DistributedJobRunner.JOB_REGISTRATION_LOCK:
+            remote: Remote = self.task.resources.node
+            job = remote.submit(
+                partial(self._run_dist_job, self.task.task_id),
+                self.task.fn,
+                self.task.args,
+                self.task.resources.gpu_ids
+            )
+            # Fast job submission from multiple threads causes lost jobs and hangs reporter thread.
+            # Keep the values above 200ms combined
+            time.sleep(0.15)
+            job.add_done_callback(_release_resource_callback)
+            time.sleep(0.15 + random.random())
+
+        return job
+
+    @staticmethod
+    def _run_dist_job(task_id, fn, args, gpu_ids):
```

Comment:
would it make sense to be a `@classmethod` so you don't need to specify DistributedJobRunner in your inner calls (`._worker`, `.__print`)?

Reply:
updated